cmd/soroban-rpc/internal/ledgerentry_storage/db.go

```diff
@@ -0,0 +1,340 @@
+package ledgerentry_storage
+
+import (
+	"context"
+	"database/sql"
+	"embed"
+	"encoding/hex"
+	"fmt"
+	"strconv"
+
+	sq "github.com/Masterminds/squirrel"
+	"github.com/jmoiron/sqlx"
+	_ "github.com/mattn/go-sqlite3"
+	migrate "github.com/rubenv/sql-migrate"
+	"github.com/stellar/go/support/errors"
+	"github.com/stellar/go/xdr"
+)
+
+//go:embed migrations/*.sql
+var migrations embed.FS
+
+var ErrEmptyDB = errors.New("DB is empty")
+
+const (
+	ledgerEntriesTableName      = "ledger_entries"
+	metaTableName               = "metadata"
+	latestLedgerSequenceMetaKey = "LatestLedgerSequence"
+)
+
+type DB interface {
+	LedgerEntryStorage
+	GetLatestLedgerSequence() (uint32, error)
+	NewLedgerEntryUpdaterTx(forLedgerSequence uint32, maxBatchSize int) (LedgerEntryUpdaterTx, error)
+}
+
+type LedgerEntryUpdaterTx interface {
+	UpsertLedgerEntry(key xdr.LedgerKey, entry xdr.LedgerEntry) error
+	DeleteLedgerEntry(key xdr.LedgerKey) error
+	Done() error
+}
+
+type sqlDB struct {
+	db                  *sqlx.DB
+	postWriteCommitHook func() error
+}
+
+func OpenSQLiteDB(dbFilePath string) (DB, error) {
+	// 1. Use Write-Ahead Logging (WAL).
+	// 2. Disable WAL auto-checkpointing (we will do the checkpointing ourselves with wal_checkpoint pragmas
+	//    after every write transaction).
+	// 3. Use synchronous=NORMAL, which is faster and still safe in WAL mode.
+	db, err := sqlx.Open("sqlite3", fmt.Sprintf("file:%s?_journal_mode=WAL&_wal_autocheckpoint=0&_synchronous=NORMAL", dbFilePath))
+	if err != nil {
+		return nil, errors.Wrap(err, "open failed")
+	}
+
+	postWriteCommitHook := func() error {
+		_, err := db.Exec("PRAGMA wal_checkpoint(TRUNCATE)")
+		return err
+	}
+
+	ret := &sqlDB{
+		db:                  db,
+		postWriteCommitHook: postWriteCommitHook,
```

Comment:
I think that implementing this method as a sqlDB method would be cleaner; i.e.
`func (db *sqlDB) PostWriteCommit() error`.

I think that the code above is not safe:
The `db` variable is a pointer; and when using anonymous functions, the compiler attempts to keep track of all the external memory location. in this case the `db` escapes the scope of the function and therefore need to be tracked. However, given that the `db` itself was not allocated as a pointer, I'm not sure if it would always work correctly.

Reply:
The garbage collector will keep track of the pointer just fine.

The idea behind using a hook was to make the implementation DB-independent (all the SQLite-specific stuff is only included in the constructor function)